runtime, implement ext_malloc  and ext_free

edwardmack opened the pull request:
<!--
 Before submitting a PR please ensure all code is commented and all tests are passing. Make sure to review the changes and ensure that only required changes are included (eg. no unnecessary reformatting by your editor)
-->
Implemented code need to use a Freeing Bump Heap Allocator for handling WASM heap via ext_malloc and ext_free functions.
<!-- Brief but specific list of changes made, describe the change in functionality rather than the change in code -->
## Changes
- Created struct and functions in `runtime/allocator.go` to implement Freeing Bump Heap Allocator to handle allocation of the WASM heap.
- Implemented 17 tests in `runtime/allocator_test.go` to replicate the tests done in the substrate allocator.
- Implemented `ext_malloc` and `ext_free` in `runtime/wasmer.go`

<!-- Details on how to run only the tests for the changes in the PR -->
## Tests:
Note: Currently some of the tests fail, so I know that there is still work needed.
```
go test -v ./runtime/...
```

<!-- Issues that this PR will close -->
<!-- 
    NOTE: you must say 'closes #xx' or 'fixes #xx' for EACH issue this closes. 
    eg: 'closes #1 and closes #2'
    See: https://help.github.com/en/articles/closing-issues-using-keywords
-->
### Fixes:
closes #191

noot commented:
looks great, I'll take a closer look soon! 

make sure to run `go fmt ./...` :)

noot commented:
I'm getting on test failing when running locally:
```
=== RUN   TestShouldAllocateMaxPossibleAllocationSize
--- FAIL: TestShouldAllocateMaxPossibleAllocationSize (0.00s)
    allocator_test.go:322: Error: allocator out of space
    allocator_test.go:326: ptr1 0
    allocator_test.go:328: Expected value of 8
```

edwardmack commented:
There are currently 2 tests failing, this is because the WASM interface we're using doesn't have functions for allocation of the heap.  I'm looking for ideas about how to deal with this.  I've details the issues on the Zube card.

edwardmack commented:
I think this is in good shape now.  It's passing tests and working.  Please review and let me know your thoughts.

noot commented on runtime/wasmer.go:
```diff
@@ -453,3 +474,14 @@ func decodeToInterface(in []byte, t interface{}) (interface{}, error) {
+var memAllocator FreeingBumpHeapAllocator
+
+// init memory allocator if it hasn't been
+func initMemAllocator(mem *wasm.Memory) {
```
I'm not sure about this method of having a global allocator, since we might eventually want multiple runtimes. we should probably put it inside the wasm context data using `instance.SetContextData(data)`

edwardmack replied:
I agree, I'd like to reference the allocator via context data, however we've discovered an issue with context data that contains pointers, this causes a `runtime error: cgo argument has Go pointer to Go pointer` error.  I can't find a way around this.

edwardmack replied:
Fixed by adding a map of RuntimeCtx to wasmer.go, so that pointer references to allocator can be stored there.

noot commented on runtime/wasmer_test.go:
```diff
@@ -651,3 +651,60 @@ func TestExt_twox_128(t *testing.T) {
 }
+
+// test ext_malloc returns expected pointer value of 8
+func TestExt_malloc(t *testing.T) {
```
maybe we can add more tests for malloc/free for various sizes?

edwardmack replied:
refactored and added tests to `allocator_test.go`

noot commented on runtime/allocator_test.go:
```diff
@@ -0,0 +1,494 @@
+		t.Fatal(err)
+	}
+	// then
+	t.Log("[TestShouldIncludePrefixesInTotalHeapSize]", "tetal_size", fbha.total_size)
```
total_size

noot commented on runtime/allocator.go:
```diff
@@ -0,0 +1,161 @@
+	// we shift 1 by three places since the first possible item size is 8
+	return 1 << 3 << index
+}
+func nextPowerOf2GT8(v uint32) uint32 {
```
add spaces between functions

edwardmack replied:
Done

edwardmack commented:
Hey all, I believe I've tackled the go pointers issues I was dealing with.  Please review and let me know your thoughts.

noot commented:
hey, I'm getting an error when running locally:
```

=== RUN   TestAllocatorShouldAllocateProperly
--- FAIL: TestAllocatorShouldAllocateProperly (0.00s)
panic: runtime error: index out of range [recovered]
	panic: runtime error: index out of range

goroutine 19 [running]:
testing.tRunner.func1(0xc000116200)
	/usr/local/go/src/testing/testing.go:830 +0x392
panic(0x8500c0, 0xe31f20)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/wasmerio/go-ext-wasm/wasmer.NewInstanceWithImports.func1(0xc00011a380, 0x11, 0x8ca6de, 0x12, 0x179c950)
	/home/elizabeth/go/pkg/mod/github.com/wasmerio/go-ext-wasm@v0.0.0-20190716093451-605a12aad995/wasmer/instance.go:90 +0x1fd
github.com/wasmerio/go-ext-wasm/wasmer.newInstanceWithImports(0xc0000ab080, 0xc000054c08, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/elizabeth/go/pkg/mod/github.com/wasmerio/go-ext-wasm@v0.0.0-20190716093451-605a12aad995/wasmer/instance.go:161 +0x307
github.com/wasmerio/go-ext-wasm/wasmer.NewInstanceWithImports(0xc000126000, 0x0, 0x200, 0xc0000ab080, 0x8e3a08, 0x7e2cb0, 0xc0000ab080, 0x0, 0x0, 0xc0000be0f0)
	/home/elizabeth/go/pkg/mod/github.com/wasmerio/go-ext-wasm@v0.0.0-20190716093451-605a12aad995/wasmer/instance.go:83 +0x76
github.com/ChainSafe/gossamer/runtime.NewRuntime(0xc0000be0f0, 0x4b, 0xc0000ab060, 0x4b, 0x0, 0x0)
	/home/elizabeth/go/src/github.com/ChainSafe/gossamer/runtime/wasmer.go:468 +0x77a
github.com/ChainSafe/gossamer/runtime.newTestRuntime(0x45faa0, 0xe, 0x25)
	/home/elizabeth/go/src/github.com/ChainSafe/gossamer/runtime/was?er_test.go:152 +0x9f
github.com/ChainSafe/gossamer/runtime.TestAllocatorShouldAllocateProperly(0xc000116200)
	/home/elizabeth/go/src/github.com/ChainSafe/gossamer/runtime/allocator_test.go:14 +0x37
testing.tRunner(0xc000116200, 0x8e3878)
	/usr/local/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x35a
FAIL	github.com/ChainSafe/gossamer/runtime	0.007s

```

but it passes on CI so maybe there's something wrong with my env? any idea?

ansermino commented on runtime/allocator.go:
```diff
@@ -0,0 +1,162 @@
+	return *fbha
+}
+
+func (fbha *FreeingBumpHeapAllocator) allocate(size uint32) (uint32, error) {
```
Please document these core functions

edwardmack replied:
Added comments.

noot commented on runtime/wasmer.go:
```diff
@@ -407,7 +470,25 @@ func NewRuntime(fp string, t *trie.Trie) (*Runtime, error) {
+		allocator: &memAllocator,
+	}
+	// add runtimeCtx to registry
+	mutex.Lock()
```
very cool solution, just wondering why we need a mutex?

edwardmack replied:
I think a mutex is needed to protect registry map from possible concurrent access.  I added more comments to the code.

edwardmack commented on runtime/allocator.go:
```diff
@@ -0,0 +1,161 @@
+// see more details at https://github.com/paritytech/substrate/issues/1615
+
+// The pointers need to be aligned to 8 bytes
+const ALIGNMENT uint32 = 8
```
Fixed, I think I've changed all places where I had snake_case instead of CamelCase.

edwardmack commented on runtime/allocator.go:
```diff
@@ -0,0 +1,161 @@
+// Creates a new allocation heap which follows a freeing-bump strategy.
+// The maximum size which can be allocated at once is 16 MiB.
+//
+// # Arguments
```
updated comments.

edwardmack replied:
Updated comments, these do seem relevant, and I updated comments to correct variable names, and that it's returning a pointer.

edwardmack commented on runtime/allocator_test.go:
```diff
@@ -0,0 +1,494 @@
+	}
+
+	// then
+	t.Log("[TestAllocatorShouldAllocateProperly]", "result", alloc_res)
```
Updated calls to Log.

edwardmack commented on runtime/allocator.go:
```diff
@@ -0,0 +1,161 @@
+	return res
+}
+
+func (fbha *FreeingBumpHeapAllocator) set_heap(ptr uint32, value uint8) {
```
updated code.

edwardmack commented on runtime/allocator_test.go:
```diff
@@ -0,0 +1,494 @@
+const PAGE_SIZE = 65536
+
+func TestAllocatorShouldAllocateProperly(t *testing.T) {
+	// given
```
I thought these were useful in tests, given this condition, when this happens, then confirm expected results.  I'm open to suggestions for commenting in testing patterns.

edwardmack commented on runtime/allocator.go:
```diff
@@ -0,0 +1,164 @@
+
+// The pointers need to be aligned to 8 bytes
+const alignment uint32 = 8
+const n = 22
```
changed `n` to `HeadsQty`

edwardmack commented on runtime/allocator.go:
```diff
@@ -0,0 +1,164 @@
+//   hence a padding might be added to align `ptr_offset` properly.
+//
+// * returns an initilized FreeingBumpHeapAllocator
+func NewAllocator(mem *wasm.Memory, ptrOffset uint32) FreeingBumpHeapAllocator {
```
Yes, updated to return a pointer instead.

edwardmack commented on runtime/allocator.go:
```diff
@@ -0,0 +1,164 @@
+	return *fbha
+}
+
+// Allocate allocates size bytes of memory from the WASM heap.
```
updated comments to better explain what's happening.

dutterbutter commented:
@edwardmack can you update branch to reflect `development` and add the license header at the beginning of `runtime/allocator.go` and any other files within the `runtime` package where it does not exist.

edwardmack commented:
Updated from development and added licenses, it's passing CI now, so it's ready for review. Let me know your thoughts.

edwardmack commented:
Hi all, I believe this code is ready now... I've updated to make the allocator part of it's own package, and updated test to work with new package.  Let me know your thoughts.
